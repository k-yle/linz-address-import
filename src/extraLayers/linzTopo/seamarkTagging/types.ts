import { SEAMARK_TYPE } from './const';

export type IHOTypes = keyof typeof SEAMARK_TYPE;
export type OSMSeamarkTypes = typeof SEAMARK_TYPE[IHOTypes];

/* eslint-disable @typescript-eslint/ban-types -- {} is okay */
type _SeamarkTypes = {
  // fields in _generic don't need to be specified in every section
  _generic: {
    fidn: string;
    /** prefer `nobjnm` */ objnam?: string;
    nobjnm?: string;
    /** prefer `ninfom` */ inform?: string;
    ninfom?: string;
    ntxtds?: string;
    /** prefer `ntxtds` */ txtdsc?: string;
    status?: string;
    sordat?: string;
    sorind?: string;
    datsta?: string;
    datend?: string;

    watlev?: string;
    colour?: string;
    colpat?: string;
    natcon?: string;
    condtn?: string;
    height?: string;
    conrad?: string;
    convis?: string;
    marsys?: string;
  };

  cable_submarine: {
    catcbl?: string;
  };
  platform: {
    verlen?: string;
    prodct?: string;
    catofp?: string;
  };
  anchorage: {
    catach?: string;
    persta?: string;
    perend?: string;
  };
  anchor_berth: {
    catach?: string;
    radius?: string;
    restrn?: string;
  };
  daymark: {
    catspm?: string;
    topshp?: string;
  };
  dumping_ground: {
    catdpg?: string;
  };
  fishing_facility: {
    catfif?: string;
  };
  fog_signal: {
    catfog?: string;
    sigper?: string;
    sigfrq?: string;
    siggrp?: string;
    sigseq?: string;
    siggen?: string;
  };
  hulk: {
    cathlk?: string;
  };
  mooring: {
    boyshp?: string;
    catmor?: string;
  };
  pile: {
    catple?: string;
  };
  pilot_boarding: {
    catpil?: string;
    // comcha?: string;
    // pildst?: string;
  };
  pontoon: {};
  pylon: {
    catpyl?: string;
  };
  radar_station: {
    catras?: string;
  };
  radar_transponder: {
    catrtb?: string;
    radwal?: string;
    siggrp?: string;
  };
  'calling-in_point': {
    comcha?: string;
    orient?: string;
    trafic?: string;
  };
  radio_station: {
    // calsgn?: string;
    catros?: string;
  };
  restricted_area: {
    catrea?: string;
    restrn?: string;
  };
  rescue_station: {
    catrsc?: string;
  };
  sand_waves: {};
  seaplane_landing_area: {};
  spring: {};
  rock: {
    expsou?: string;
    natsur?: string;
    quasou?: string;
    valsou?: string;
    tecsou?: string;
    souacc?: string;
  };
  weed: {
    catwed?: string;
  };
  topmark: {
    catspm?: string;
    topshp?: string;
  };
  light: {
    catlit?: string;
    litchr?: string;
    litvis?: string;
    sectr1?: string;
    sectr2?: string;
    siggrp?: string;
    sigper?: string;
    sigseq?: string;
    valnmr?: string;
    mltylt?: string;
    exclit?: string;
  };
  distance_mark: {
    catdis?: string;
  };
  obstruction: {
    catobs?: string;
    tecsou?: string;
    natqua?: string;
  };
  beacon_cardinal: {
    catcam?: string;
    bcnshp?: string;
  };
  beacon_isolated_danger: {
    bcnshp?: string;
  };
  beacon_lateral: {
    bcnshp?: string;
    catlam?: string;
  };
  beacon_safe_water: {
    bcnshp?: string;
  };
  beacon_special_purpose: {
    bcnshp?: string;
    catspm?: string;
    elevat?: string;
  };
  buoy_cardinal: {
    boyshp?: string;
    catcam?: string;
  };
  buoy_isolated_danger: {
    boyshp?: string;
  };
  buoy_lateral: {
    boyshp?: string;
    catlam?: string;
  };
  buoy_safe_water: {
    boyshp?: string;
  };
  buoy_special_purpose: {
    boyshp?: string;
    catspm?: string;
    veracc?: string; // no seamark tag
    verlen?: string;
    picrep?: string;
  };
  wreck: {
    catwrk?: string;
  };
};
/* eslint-enable @typescript-eslint/ban-types */

// if this fails, one of the keys in the map above is invalid
// eslint-disable-next-line @typescript-eslint/no-unused-vars
type CompileTest = Pick<
  { [P in OSMSeamarkTypes]: never },
  keyof Omit<_SeamarkTypes, '_generic'>
>;

// from https://stackoverflow.com/a/50375286/5470183
type UnionToIntersection<U> = (
  U extends unknown ? (k: U) => void : never
) extends (k: infer I) => void
  ? I
  : never;

/**
 * one huge type with every property defined in IHO-S57
 *
 * typescript is even smart enough to merge jsdoc comments for each attr
 */
export type AllSeamarkProps = UnionToIntersection<
  _SeamarkTypes[keyof _SeamarkTypes]
>;

// a bit of magic to merge the "_generic" object into each object, then delete "_generic"
export type Seamark = Omit<
  { [P in keyof _SeamarkTypes]: _SeamarkTypes[P] & _SeamarkTypes['_generic'] },
  '_generic'
>;
