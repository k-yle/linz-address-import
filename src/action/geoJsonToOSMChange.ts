import {
  GeoJson,
  GeoJsonPoint,
  GeoJsonLine,
  GeoJsonMultiPolygon,
} from '../types';

type Tags = Record<string, string | undefined>;

/** when you access `ID.node` etc., it will always give you a new ID */
class IDManager {
  #count = { node: 0, way: 0, relation: 0 };

  get node() {
    this.#count.node -= 1;
    return this.#count.node;
  }

  get way() {
    this.#count.way -= 1;
    return this.#count.way;
  }

  get relation() {
    this.#count.relation -= 1;
    return this.#count.relation;
  }

  get changesetSize() {
    return -this.#count.node + -this.#count.way + -this.#count.relation;
  }
}

const tagsToXml = (tags: Tags) =>
  Object.entries(tags)
    .filter(([k, v]) => k && v)
    .map(([k, v]) => `<tag k="${k}" v="${v}" />`)
    .join('\n      ');

function handleNode(ID: IDManager, geometry: GeoJsonPoint, tags: Tags) {
  const [lng, lat] = geometry.coordinates;
  return `
    <node id="${ID.node}" lon="${lng}" lat="${lat}" version="0">
      ${tagsToXml(tags)}
    </node>`;
}

function handleWay(ID: IDManager, geometry: GeoJsonLine, tags: Tags) {
  let str = '';
  const nodes = [];
  for (const [lng, lat] of geometry.coordinates) {
    const id = ID.node;
    nodes.push(id);
    str += `<node id="${id}" lon="${lng}" lat="${lat}" version="0" />\n    `;
  }

  str += `
    <way id="${ID.way}" version="0">
      ${nodes.map((id) => `<nd ref="${id}" />`).join('\n      ')}
      ${tagsToXml(tags)}
    </way>
  `;

  return str;
}

function handleMultiPolygon(
  ID: IDManager,
  geometry: GeoJsonMultiPolygon,
  tags: Tags,
) {
  let str = '';
  const ways: { wayId: number; role: string }[] = [];
  for (const group of geometry.coordinates) {
    for (let i = 0; i < group.length; i += 1) {
      const shape = group[i];
      const nodes = [];
      for (const [lng, lat] of shape) {
        const nodeId = ID.node;
        nodes.push(nodeId);
        str += `<node id="${nodeId}" lon="${lng}" lat="${lat}" version="0" />\n    `;
      }
      const wayId = ID.way;
      ways.push({ wayId, role: i === 0 ? 'outer' : 'inner' });
      str += `
        <way id="${wayId}" version="0">
          ${nodes.map((id) => `<nd ref="${id}" />`).join('\n    ')}
        </way>
      `;
    }
  }
  str += `
    <relation id="${ID.relation}" version="0">
      ${ways
        .map((w) => `<member type="way" role="${w.role}" ref="${w.wayId}"/>`)
        .join('\n     ')}
      ${tagsToXml(tags)}
    </relation>
  `;
  return str;
}

export function geoJsonToOSMChange(
  geoJson: GeoJson,
  fileName: string,
  count: string,
): { osmChange: string; tooBig: boolean } {
  const ID = new IDManager();

  let out = '';
  for (const feature of geoJson.features) {
    switch (feature.geometry.type) {
      case 'Point': {
        out += handleNode(ID, feature.geometry, feature.properties);
        break;
      }

      case 'LineString': {
        // way
        out += handleWay(ID, feature.geometry, feature.properties);
        break;
      }

      case 'Polygon': {
        if (feature.geometry.coordinates.length === 1) {
          // area
          // to simplify the code, we can treat a geojson polygon with only one outer ring as a (closed)-way
          const geom: GeoJsonLine = {
            type: 'LineString',
            coordinates: feature.geometry.coordinates[0],
          };
          out += handleWay(ID, geom, feature.properties);
        } else {
          // multipolygon
          // a geojson polygon with only multiple outer rings is just a simple geojson multipolygon
          const geom: GeoJsonMultiPolygon = {
            type: 'MultiPolygon',
            coordinates: [feature.geometry.coordinates],
          };
          out += handleMultiPolygon(ID, geom, feature.properties);
        }
        break;
      }

      case 'MultiPolygon': {
        out += handleMultiPolygon(ID, feature.geometry, feature.properties);
        break;
      }

      // will never happen
      default:
        throw new Error(
          // @ts-expect-error ts knows this can never happen
          `Invalid geojson type in ${fileName}: ${feature.geometry.type}`,
        );
    }
  }

  const tooBig = ID.changesetSize >= 10_000;

  // this would brek the snapshot tests
  const date =
    process.env.NODE_ENV === 'test'
      ? 'DATE_GOES_HERE'
      : new Date().toISOString();

  const osmChange = `
<osmChange version="0.6" generator="LINZ Data Import" LINZ_title="${fileName}" LINZ_count="${count}" LINZ_generated_on="${date}">
  <create>
    ${out}
  </create>
  <modify />
  <delete if-unused="true" />
</osmChange>`;
  return { osmChange, tooBig };
}
