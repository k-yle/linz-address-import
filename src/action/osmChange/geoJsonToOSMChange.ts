import {
  OsmChange,
  OsmFeature,
  OsmNode,
  OsmRelation,
  OsmWay,
  createOsmChangeXml,
} from 'osm-api';
import { MAP } from '../util';
import {
  GeoJson,
  GeoJsonPoint,
  GeoJsonLine,
  GeoJsonMultiPolygon,
} from '../../types';
import { amendOsmChangeFeature } from './amendOsmChangeFeature';

type Tags = Record<string, string | undefined>;

/** when you access `ID.node` etc., it will always give you a new ID */
class IDManager {
  #count = { node: 0, way: 0, relation: 0 };

  get node() {
    this.#count.node -= 1;
    return this.#count.node;
  }

  get way() {
    this.#count.way -= 1;
    return this.#count.way;
  }

  get relation() {
    this.#count.relation -= 1;
    return this.#count.relation;
  }

  get changesetSize() {
    return -this.#count.node + -this.#count.way + -this.#count.relation;
  }
}

function handleNode(
  ID: IDManager,
  geometry: GeoJsonPoint,
  tags: Tags,
): OsmFeature[] {
  const [lon, lat] = geometry.coordinates;
  const node = { type: 'node', id: ID.node, lat, lon, tags } as OsmNode;

  return [node];
}

function handleWay(
  ID: IDManager,
  geometry: GeoJsonLine,
  tags: Tags,
): OsmFeature[] {
  const nodes: OsmNode[] = [];
  for (const [lon, lat] of geometry.coordinates) {
    const node = { type: 'node', id: ID.node, lat, lon } as OsmNode;
    nodes.push(node);
  }

  const way = {
    type: 'way',
    id: ID.way,
    tags,
    nodes: nodes.map((n) => n.id),
  } as OsmWay;

  return [...nodes, way];
}

function handleMultiPolygon(
  ID: IDManager,
  geometry: GeoJsonMultiPolygon,
  tags: Tags,
): OsmFeature[] {
  const nodes: OsmNode[] = [];
  const ways: OsmWay[] = [];
  const members: OsmRelation['members'] = [];

  for (const group of geometry.coordinates) {
    for (let i = 0; i < group.length; i += 1) {
      const shape = group[i];
      const theseNodes: OsmNode[] = [];

      for (const [lon, lat] of shape) {
        const node = { type: 'node', id: ID.node, lat, lon } as OsmNode;
        theseNodes.push(node);
      }

      const way = {
        type: 'way',
        id: ID.way,
        nodes: theseNodes.map((n) => n.id),
      } as OsmWay;
      ways.push(way);
      members.push({
        type: 'way',
        ref: way.id,
        role: i === 0 ? 'outer' : 'inner',
      });
    }
  }

  const relation = {
    type: 'relation',
    id: ID.relation,
    tags,
    members,
  } as OsmRelation;

  return [...nodes, ...ways, relation];
}

export function geoJsonToOSMChange(
  geoJson: GeoJson,
  fileName: string,
  count: string,
): { osmChange: string; tooBig: boolean } {
  const ID = new IDManager();

  const osmChange: OsmChange = { create: [], modify: [], delete: [] };

  for (const feature of geoJson.features) {
    if (feature.id.startsWith('SPECIAL_EDIT_')) {
      // you can edit nodes, ways, and relations
      const { __osmId, ...tags } = feature.properties;
      const id = +__osmId!.slice(1);
      const type = MAP[__osmId![0] as 'n' | 'w' | 'r'];

      const editedOsmFeat = amendOsmChangeFeature(type, id, tags);

      osmChange.modify.push(editedOsmFeat);
    } else if (feature.id.startsWith('SPECIAL_DELETE_')) {
      // you can only delete nodes
      const [lon, lat] = (feature.geometry as GeoJsonPoint).coordinates;
      const { __osmId, ...tags } = feature.properties;
      const node = {
        type: 'node',
        lat,
        lon,
        id: +__osmId!.slice(1),
        tags,
      } as OsmNode;
      osmChange.delete.push(node);
    } else if (feature.id.startsWith('SPECIAL_MOVE_')) {
      // you can only move nodes
      const [lon, lat] = (feature.geometry as GeoJsonLine).coordinates[1];
      const { __osmId, ...tags } = feature.properties;
      const node = {
        type: 'node',
        lat,
        lon,
        id: +__osmId!.slice(1),
        tags,
      } as OsmNode;
      osmChange.modify.push(node);
    } else {
      // create
      switch (feature.geometry.type) {
        case 'Point': {
          osmChange.create.push(
            ...handleNode(ID, feature.geometry, feature.properties),
          );
          break;
        }

        case 'LineString': {
          // way
          osmChange.create.push(
            ...handleWay(ID, feature.geometry, feature.properties),
          );
          break;
        }

        case 'Polygon': {
          if (feature.geometry.coordinates.length === 1) {
            // area
            // to simplify the code, we can treat a geojson polygon with only one outer ring as a (closed)-way
            const geom: GeoJsonLine = {
              type: 'LineString',
              coordinates: feature.geometry.coordinates[0],
            };
            osmChange.create.push(...handleWay(ID, geom, feature.properties));
          } else {
            // multipolygon
            // a geojson polygon with only multiple outer rings is just a simple geojson multipolygon
            const geom: GeoJsonMultiPolygon = {
              type: 'MultiPolygon',
              coordinates: [feature.geometry.coordinates],
            };
            osmChange.create.push(
              ...handleMultiPolygon(ID, geom, feature.properties),
            );
          }
          break;
        }

        case 'MultiPolygon': {
          osmChange.create.push(
            ...handleMultiPolygon(ID, feature.geometry, feature.properties),
          );
          break;
        }

        // will never happen
        default:
          throw new Error(
            // @ts-expect-error ts knows this can never happen
            `Invalid geojson type in ${fileName}: ${feature.geometry.type}`,
          );
      }
    }
  }

  const tooBig = ID.changesetSize >= 10_000;

  // this would brek the snapshot tests
  const date =
    process.env.NODE_ENV === 'test'
      ? 'DATE_GOES_HERE'
      : new Date().toISOString();

  const xml = createOsmChangeXml(0, osmChange, {
    $generator: 'LINZ Data Import',
    $LINZ_title: fileName,
    $LINZ_count: count,
    $LINZ_generated_on: date,
  });
  return { osmChange: xml, tooBig };
}
