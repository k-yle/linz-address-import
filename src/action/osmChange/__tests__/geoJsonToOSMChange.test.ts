import { GeoJson, GeoJsonFeature } from '../../../types';
import { geoJsonToOSMChange } from '../geoJsonToOSMChange';

const geoJson = (features: GeoJsonFeature[]): GeoJson => ({
  type: 'FeatureCollection',
  features,
});

const Point: GeoJsonFeature = {
  type: 'Feature',
  id: '1',
  geometry: {
    type: 'Point',
    coordinates: [174.7, -36.9],
  },
  properties: {
    natural: 'peak',
    name: undefined,
    description: '',
  },
};
const LineString: GeoJsonFeature = {
  type: 'Feature',
  id: '2',
  geometry: {
    type: 'LineString',
    coordinates: [
      [174.5, -37],
      [174.6, -37],
    ],
  },
  properties: { highway: 'footway' },
};

const Polygon: GeoJsonFeature = {
  type: 'Feature',
  id: '3',
  geometry: {
    type: 'Polygon',
    coordinates: [
      [
        [174.8, -36.9],
        [174.81, -36.91],
        [174.82, -36.92],
        [174.8, -36.9],
      ],
    ],
  },
  properties: {
    ref: '1',
    landuse: 'aquaculture',
  },
};
const MultiPolygon: GeoJsonFeature = {
  type: 'Feature',
  id: '4',
  geometry: {
    type: 'MultiPolygon',
    coordinates: [
      [
        // outer way 1:
        [
          [1, 2],
          [3, 4],
          [5, 6],
          [7, 8],
        ],
        // inner way 1:
        [
          [15, 16],
          [17, 18],
          [19, 20],
          [21, 22],
        ],
        // inner way 2:
        [
          [23, 24],
          [25, 26],
          [27, 28],
        ],
      ],
      [
        // outer way 2:
        [
          [9, 10],
          [11, 12],
          [13, 14],
        ],
      ],
    ],
  },
  properties: {
    ref: '1',
    landuse: 'aquaculture',
  },
};

describe('geoJsonToOSMChange', () => {
  it('can convert files with Point', () => {
    const file = geoJson([Point]);
    expect(
      geoJsonToOSMChange(file, 'test layer', 'add 1').osmChange,
    ).toMatchSnapshot();
  });

  it('can convert files with LineString', () => {
    const file = geoJson([LineString]);
    expect(
      geoJsonToOSMChange(file, 'test layer', 'add 1').osmChange,
    ).toMatchSnapshot();
  });

  it('can convert files with Polygon', () => {
    const file = geoJson([Polygon]);
    expect(
      geoJsonToOSMChange(file, 'test layer', 'add 1').osmChange,
    ).toMatchSnapshot();
  });

  it('can convert files with MultiPolygon', () => {
    const file = geoJson([MultiPolygon]);
    expect(
      geoJsonToOSMChange(file, 'test layer', 'add 1').osmChange,
    ).toMatchSnapshot();
  });

  it('can convert files with all features', () => {
    const file = geoJson([Point, LineString, Polygon, LineString]);
    expect(
      geoJsonToOSMChange(file, 'test layer', 'add 4').osmChange,
    ).toMatchSnapshot();
  });
});
