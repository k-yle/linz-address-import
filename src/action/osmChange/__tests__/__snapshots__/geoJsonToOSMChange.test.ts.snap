// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geoJsonToOSMChange can convert files with LineString 1`] = `
"<osmChange version=\\"0.6\\" generator=\\"LINZ Data Import\\" LINZ_title=\\"test layer\\" LINZ_count=\\"add 1\\" LINZ_generated_on=\\"DATE_GOES_HERE\\">
  <create>
    <node id=\\"-1\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-37\\" lon=\\"174.5\\"/>
    <node id=\\"-2\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-37\\" lon=\\"174.6\\"/>
    <way id=\\"-1\\" version=\\"0\\" changeset=\\"0\\">
      <tag k=\\"highway\\" v=\\"footway\\"/>
      <nd ref=\\"-1\\"/>
      <nd ref=\\"-2\\"/>
    </way>
  </create>
  <modify/>
  <delete if-unused=\\"true\\"/>
</osmChange>
"
`;

exports[`geoJsonToOSMChange can convert files with MultiPolygon 1`] = `
"<osmChange version=\\"0.6\\" generator=\\"LINZ Data Import\\" LINZ_title=\\"test layer\\" LINZ_count=\\"add 1\\" LINZ_generated_on=\\"DATE_GOES_HERE\\">
  <create>
    <way id=\\"-1\\" version=\\"0\\" changeset=\\"0\\">
      <nd ref=\\"-1\\"/>
      <nd ref=\\"-2\\"/>
      <nd ref=\\"-3\\"/>
      <nd ref=\\"-4\\"/>
    </way>
    <way id=\\"-2\\" version=\\"0\\" changeset=\\"0\\">
      <nd ref=\\"-5\\"/>
      <nd ref=\\"-6\\"/>
      <nd ref=\\"-7\\"/>
      <nd ref=\\"-8\\"/>
    </way>
    <way id=\\"-3\\" version=\\"0\\" changeset=\\"0\\">
      <nd ref=\\"-9\\"/>
      <nd ref=\\"-10\\"/>
      <nd ref=\\"-11\\"/>
    </way>
    <way id=\\"-4\\" version=\\"0\\" changeset=\\"0\\">
      <nd ref=\\"-12\\"/>
      <nd ref=\\"-13\\"/>
      <nd ref=\\"-14\\"/>
    </way>
    <relation id=\\"-1\\" version=\\"0\\" changeset=\\"0\\">
      <tag k=\\"ref\\" v=\\"1\\"/>
      <tag k=\\"landuse\\" v=\\"aquaculture\\"/>
      <member type=\\"way\\" ref=\\"-1\\" role=\\"outer\\"/>
      <member type=\\"way\\" ref=\\"-2\\" role=\\"inner\\"/>
      <member type=\\"way\\" ref=\\"-3\\" role=\\"inner\\"/>
      <member type=\\"way\\" ref=\\"-4\\" role=\\"outer\\"/>
    </relation>
  </create>
  <modify/>
  <delete if-unused=\\"true\\"/>
</osmChange>
"
`;

exports[`geoJsonToOSMChange can convert files with Point 1`] = `
"<osmChange version=\\"0.6\\" generator=\\"LINZ Data Import\\" LINZ_title=\\"test layer\\" LINZ_count=\\"add 1\\" LINZ_generated_on=\\"DATE_GOES_HERE\\">
  <create>
    <node id=\\"-1\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.9\\" lon=\\"174.7\\">
      <tag k=\\"natural\\" v=\\"peak\\"/>
      <tag k=\\"name\\"/>
      <tag k=\\"description\\" v=\\"\\"/>
    </node>
  </create>
  <modify/>
  <delete if-unused=\\"true\\"/>
</osmChange>
"
`;

exports[`geoJsonToOSMChange can convert files with Polygon 1`] = `
"<osmChange version=\\"0.6\\" generator=\\"LINZ Data Import\\" LINZ_title=\\"test layer\\" LINZ_count=\\"add 1\\" LINZ_generated_on=\\"DATE_GOES_HERE\\">
  <create>
    <node id=\\"-1\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.9\\" lon=\\"174.8\\"/>
    <node id=\\"-2\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.91\\" lon=\\"174.81\\"/>
    <node id=\\"-3\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.92\\" lon=\\"174.82\\"/>
    <node id=\\"-4\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.9\\" lon=\\"174.8\\"/>
    <way id=\\"-1\\" version=\\"0\\" changeset=\\"0\\">
      <tag k=\\"ref\\" v=\\"1\\"/>
      <tag k=\\"landuse\\" v=\\"aquaculture\\"/>
      <nd ref=\\"-1\\"/>
      <nd ref=\\"-2\\"/>
      <nd ref=\\"-3\\"/>
      <nd ref=\\"-4\\"/>
    </way>
  </create>
  <modify/>
  <delete if-unused=\\"true\\"/>
</osmChange>
"
`;

exports[`geoJsonToOSMChange can convert files with all features 1`] = `
"<osmChange version=\\"0.6\\" generator=\\"LINZ Data Import\\" LINZ_title=\\"test layer\\" LINZ_count=\\"add 4\\" LINZ_generated_on=\\"DATE_GOES_HERE\\">
  <create>
    <node id=\\"-1\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.9\\" lon=\\"174.7\\">
      <tag k=\\"natural\\" v=\\"peak\\"/>
      <tag k=\\"name\\"/>
      <tag k=\\"description\\" v=\\"\\"/>
    </node>
    <node id=\\"-2\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-37\\" lon=\\"174.5\\"/>
    <node id=\\"-3\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-37\\" lon=\\"174.6\\"/>
    <node id=\\"-4\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.9\\" lon=\\"174.8\\"/>
    <node id=\\"-5\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.91\\" lon=\\"174.81\\"/>
    <node id=\\"-6\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.92\\" lon=\\"174.82\\"/>
    <node id=\\"-7\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-36.9\\" lon=\\"174.8\\"/>
    <node id=\\"-8\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-37\\" lon=\\"174.5\\"/>
    <node id=\\"-9\\" version=\\"0\\" changeset=\\"0\\" lat=\\"-37\\" lon=\\"174.6\\"/>
    <way id=\\"-1\\" version=\\"0\\" changeset=\\"0\\">
      <tag k=\\"highway\\" v=\\"footway\\"/>
      <nd ref=\\"-2\\"/>
      <nd ref=\\"-3\\"/>
    </way>
    <way id=\\"-2\\" version=\\"0\\" changeset=\\"0\\">
      <tag k=\\"ref\\" v=\\"1\\"/>
      <tag k=\\"landuse\\" v=\\"aquaculture\\"/>
      <nd ref=\\"-4\\"/>
      <nd ref=\\"-5\\"/>
      <nd ref=\\"-6\\"/>
      <nd ref=\\"-7\\"/>
    </way>
    <way id=\\"-3\\" version=\\"0\\" changeset=\\"0\\">
      <tag k=\\"highway\\" v=\\"footway\\"/>
      <nd ref=\\"-8\\"/>
      <nd ref=\\"-9\\"/>
    </way>
  </create>
  <modify/>
  <delete if-unused=\\"true\\"/>
</osmChange>
"
`;
